### ISSUE TITLE
Consider adding __repr__ methods to all classes

### ISSUE THREAD

**Issue by matthewfeickert** [diana-hep/pyhf#476]
# Description

Describe the feature you would like to be added and its use. Please give a clear and concise description.

For future debugging instances (I'm thinking more of users who are debugging rather but also for the core devs) it could be quite useful to have [`__repr__`](https://docs.python.org/3/library/functions.html#repr) methods for all relevant classes that return something simple but useful such as the string representation of the constructor call for that class with the `__init__` arguments.

So perhaps something along the lines of

```python
def __repr__(self, kwargs):
    return '{}(self.attribute_1, {})'.format(cls.__name__, ', '.join(str(k)+'='+str(v) for k,v in kwargs.items()))
```

**Comment by kratsg** [comment]
Why doesn't `help(ExampleClass)` do the trick? I think in general, it's not pythonic to override the `__repr__` methods unless you've got a very good reason.

**Comment by matthewfeickert** [comment]
> Why doesn't `help(ExampleClass)` do the trick? 

This is fine if you have a general question about the class, but I think while debugging if you want to be able to get information on a particular class instance and know how _it itself_ was constructed this won't be helpful.

> I think in general, it's not pythonic to override the `__repr__` methods unless you've got a very good reason.

It seems that many libraries do this though. Check out matplotlib for example (NumPy too).

```python
>>> import matplotlib.pyplot as plt
>>> fig, ax = plt.subplots(1, 1)
>>> repr(fig)
'<Figure size 640x480 with 1 Axes>'
>>> str(fig)
'Figure(640x480)'
>>> class ExampleClass:
...     def __init__(self, arg_1, **kwargs):
...             self.attribute_1 = arg_1
...             self.kwargs = kwargs
...
>>> example = ExampleClass(10, test=20)
>>> repr(example)
'<__main__.ExampleClass object at 0x7fd85abca2b0>'
>>> str(example)
'<__main__.ExampleClass object at 0x7fd85abca2b0>'
```

**Comment by kratsg** [comment]
> ```python
> >>> repr(fig)
> '<Figure size 640x480 with 1 Axes>'
> ```

This is not a good habit, since you lose information about the underlying memory pointer... (e.g. https://codingkilledthecat.wordpress.com/2012/06/03/please-dont-abuse-repr/). If we can make it as clear as possible, then it should be fine.

**Comment by lukasheinrich** [comment]
Agree I like pretty printed repr but we should keep the pointer

**Comment by matthewfeickert** [comment]
> This is not a good habit, since you lose information about the underlying memory pointer.

> Agree I like pretty printed repr but we should keep the pointer

Yes, I very much agree. If you look at my `ExampleClass` example I already have this implemented (my motivation for this is debugging help and without the pointer it isn't worth it).

```python
>>> example = ExampleClass(10, test=20)
>>> print(example)
ExampleClass(10, test=20) at 0x7fa6d98aa5f8
>>> print(repr(example))
ExampleClass(10, test=20) at 0x7fa6d98aa5f8
```

**Comment by kratsg** [comment]
I'm ok with this proposal.

**Comment by matthewfeickert** [comment]
> I'm ok with this proposal.

Now I just need to finish a thesis so I have time to implement this. :)
